swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java

```diff
@@ -88,18 +93,44 @@ public Double getContractCoverage(BillableUsage usage) throws ContractMissingExc
       throw new ContractMissingException(
           String.format("No contract info found for usage! %s", usage));
     }
-    Integer totalUnderContract =
-        contracts.stream()
-            .filter(contract -> isValidContract(contract, usage))
-            .map(
-                c ->
-                    c.getMetrics().stream()
-                        .filter(metric -> metric.getMetricId().equals(contractMetricId))
-                        .map(Metric::getValue)
-                        .reduce(0, Integer::sum))
-            .reduce(0, Integer::sum);
-    log.debug("Total contract coverage is {} for usage {} ", totalUnderContract, usage);
-    return Double.valueOf(totalUnderContract);
+
+    boolean hasGratisStatus =
+        SubscriptionDefinition.isMetricGratis(usage.getProductId(), usage.getMetricId());
+    ContractCoverage coverage = new ContractCoverage(contractMetricId);
+    for (Contract contract : contracts) {
+      if (isValidContract(contract, usage)) {
+        var value =
+            contract.getMetrics().stream()
+                .filter(metric -> metric.getMetricId().equals(contractMetricId))
+                .map(Metric::getValue)
+                .reduce(0, Integer::sum);
+        if (hasGratisStatus && isGratisContract(contract)) {
+          // Directly return that the coverage is gratis.
+          coverage.setGratis(true);
+          return coverage;
+        } else {
```

Comment:
As mentioned elsewhere, rather than short-circuiting, the coverage should be considered gratis only when all matching contracts are gratis. So it's probably easiest to store vars for `hasGratisContracts` and `hasNonGratisContracts` and then set `coverage.setGratis(hasGratisContracts && !hasNonGratisContracts)` outside of the loop (or something equivalent).

Reply:
thanks for clarifying, will change it